### PR TITLE
SSE C support

### PR DESCRIPTION
**Pull request by mpnowacki-reef**

### CODE REVIEW

File: test/integration/test_b2_command_line.py [hunk 1]
```diff
@@ -876,23 +932,51 @@ def sync_down_helper(b2_tool, bucket_name, folder_in_bucket):
...
+            'B2_DESTINATION_SSE_C_KEY_B64': base64.b64encode(encryption.key.secret).decode(),
+            'B2_DESTINATION_SSE_C_KEY_ID': encryption.key.key_id,
+        }
+        sync_encryption_args = ['--sourceServerSideEncryption', 'SSE-C']
+        sync_additional_env = {
+            'B2_SOURCE_SSE_C_KEY_B64': base64.b64encode(encryption.key.secret).decode(),
+            'B2_SOURCE_SSE_C_KEY_ID': encryption.key.key_id,
+        }
+    else:
+        raise ValueError(encryption)
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:952]:
```suggestion
        raise NotImplementedError(encryption)
```

File: test/integration/test_b2_command_line.py [hunk 2]
```diff
@@ -876,23 +932,51 @@ def sync_down_helper(b2_tool, bucket_name, folder_in_bucket):
...
         )
         b2_tool.should_succeed(
-            ['upload-file', '--noProgress', bucket_name, file_to_upload, b2_file_prefix + 'b']
+            ['upload-file', '--noProgress', bucket_name, file_to_upload, b2_file_prefix + 'b'] +
+            upload_encryption_args,
+            additional_env=upload_additional_env,
+        )
+        b2_tool.should_succeed(
+            ['sync', b2_sync_point, local_path] + sync_encryption_args,
+            additional_env=sync_additional_env
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:972]:
```suggestion
            additional_env=sync_additional_env,
```

File: test/integration/test_b2_command_line.py [hunk 3]
```diff
@@ -947,40 +1073,98 @@ def sync_copy_helper(
...
+                ],
+                additional_env={
+                    'B2_DESTINATION_SSE_C_KEY_B64':
+                        base64.b64encode(source_encryption.key.secret).decode(),
+                    'B2_DESTINATION_SSE_C_KEY_ID':
+                        source_encryption.key.key_id,
+                },
+            )
+    else:
+        raise ValueError(source_encryption)
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:1103]:
```suggestion
        raise NotImplementedError(source_encryption)
```

File: test/integration/test_b2_command_line.py [hunk 4]
```diff
@@ -947,40 +1073,98 @@ def sync_copy_helper(
...
+            b2_tool.should_succeed(
+                [
+                    'upload-file', '--noProgress', '--destinationServerSideEncryption', 'SSE-C',
+                    bucket_name, file_to_upload, b2_file_prefix + suffix
+                ],
+                additional_env={
+                    'B2_DESTINATION_SSE_C_KEY_B64':
+                        base64.b64encode(source_encryption.key.secret).decode(),
+                    'B2_DESTINATION_SSE_C_KEY_ID':
+                        source_encryption.key.key_id,
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:1099]: I think sdk should warn if encryption key ids are different (whether decryption works or not)

File: test/integration/test_b2_command_line.py [hunk 5]
```diff
@@ -947,40 +1073,98 @@ def sync_copy_helper(
...
+                    source_encryption.key.key_id,
+            }
+        )
+        expected_encryption_str = encryption_summary(
+            expected_encryption.as_dict(), {'sse_c_key_id': destination_encryption.key.key_id}
+        )
+
     else:
-        raise ValueError(expected_encryption)
+        raise ValueError(destination_encryption)
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:1164]:
```suggestion
        raise NotImplementedError(destination_encryption)
```

File: test/integration/test_b2_command_line.py [hunk 6]
```diff
@@ -1122,24 +1306,226 @@ def sse_b2_test(b2_tool, bucket_name):
     #should_equal({'mode': 'none'}, file_info['serverSideEncryption'])  # v1 B2Api.get_file_info is a legacy interface which returns raw server response
 
 
+def sse_c_test(b2_tool, bucket_name):
+
+    file_to_upload = 'README.md'
+    secret = os.urandom(32)
```
Comment (ppolewicz) [inline test/integration/test_b2_command_line.py:1312]: import secrets again

File: b2/console_tool.py [hunk 7]
```diff
@@ -196,12 +235,61 @@ def _get_destination_sse_setting(cls, args):
...
-            return EncryptionSetting(mode=mode, algorithm=algorithm)
+            key = None
+            if mode == EncryptionMode.SSE_C:
+                encryption_key_b64 = os.environ.get(B2_DESTINATION_SSE_C_KEY_B64_ENV_VAR)
+                if not encryption_key_b64:
+                    raise ValueError(
+                        'Using SSE-C requires providing an encryption key via %s env var' %
+                        B2_DESTINATION_SSE_C_KEY_B64_ENV_VAR
+                    )
+                key_id = os.environ.get(B2_DESTINATION_SSE_C_KEY_ID_ENV_VAR)
```
[the remark this answers is not in the document]
Comment (mpnowacki-reef) [reply b2/console_tool.py:246]: A warning will be nice, it should be issued in CLI - in SDK you always have to explicitly provide `EncryptionKey.key_id`, even if it's `None`, so the warning there is not necessary

File: test/integration/test_b2_command_line.py (hunk 4 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:1099]: SDK actually throws an error in this case

File: test/integration/test_b2_command_line.py (hunk 6 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:1312]: the `secrets` module was introduced in 3.6

Comment (ppolewicz) [reply test/integration/test_b2_command_line.py:1316]: 🤦

**Pull request merged**